File: src/main/java/gargoyle/netsquares/prop/simple/EnumProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.RWProperty;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.Objects;

public final class EnumProperty<E extends Enum<E>> extends RWProperty<E> implements Comparable<EnumProperty<E>>, Serializable {
    public EnumProperty(E value) {
        super(value);
    }

    @Override
    public int compareTo(@NotNull EnumProperty<E> o) {
        return Integer.compare(get().ordinal(), o.get().ordinal());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(get(), that.get());
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/EnumProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.RWProperty

class EnumProperty<E : Enum<E>>(value: E) : RWProperty<E>(value), Comparable<EnumProperty<E>> {
    override fun compareTo(other: EnumProperty<E>): Int {
        return get().ordinal.compareTo(other.get().ordinal)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return get() == other.get()
    }

    override fun hashCode(): Int = get().hashCode()
}
